dbms/src/Interpreters/ActionsVisitor.cpp

```diff
@@ -107,37 +93,77 @@ SetPtr makeExplicitSet(
         return 0;
     };
 
-    size_t left_tuple_depth = getTupleDepth(left_arg_type);
-    size_t right_tuple_depth = getTupleDepth(right_arg_type);
+    const auto& right_arg_type = right_arg_evaluated.second;
+    const auto& right_arg_value = right_arg_evaluated.first;
 
-    ASTPtr elements_ast = nullptr;
-
-    /// 1 in 1; (1, 2) in (1, 2); identity(tuple(tuple(tuple(1)))) in tuple(tuple(tuple(1))); etc.
-    if (left_tuple_depth == right_tuple_depth)
+    const size_t left_tuple_depth = getTupleDepth(left_arg_type);
+    const size_t right_tuple_depth = getTupleDepth(right_arg_type);
+    if (left_tuple_depth != right_tuple_depth && left_tuple_depth + 1 != right_tuple_depth)
     {
-        ASTPtr exp_list = std::make_shared<ASTExpressionList>();
-        exp_list->children.push_back(right_arg);
-        elements_ast = exp_list;
+        throw Exception("Invalid types for IN function: "
+                        + left_arg_type->getName() + " and " + right_arg_type->getName() + ".",
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
     }
-    /// 1 in (1, 2); (1, 2) in ((1, 2), (3, 4)); etc.
-    else if (left_tuple_depth + 1 == right_tuple_depth)
+
+    Block block;
+    auto col = left_arg_type->createColumn();
+    switch (right_arg_type->getTypeId())
     {
-        const auto * set_func = right_arg->as<ASTFunction>();
+        case TypeIndex::UInt8: [[fallthrough]];
+        case TypeIndex::UInt16: [[fallthrough]];
+        case TypeIndex::UInt32: [[fallthrough]];
+        case TypeIndex::UInt64: [[fallthrough]];
+        case TypeIndex::UInt128: [[fallthrough]];
+        case TypeIndex::Int8: [[fallthrough]];
+        case TypeIndex::Int16: [[fallthrough]];
+        case TypeIndex::Int32: [[fallthrough]];
+        case TypeIndex::Int64: [[fallthrough]];
+        case TypeIndex::Int128: [[fallthrough]];
+        case TypeIndex::Float32: [[fallthrough]];
+        case TypeIndex::Float64: [[fallthrough]];
+        case TypeIndex::Date: [[fallthrough]];
+        case TypeIndex::DateTime: [[fallthrough]];
+        case TypeIndex::String: [[fallthrough]];
+        case TypeIndex::FixedString: [[fallthrough]];
+        case TypeIndex::Enum8: [[fallthrough]];
+        case TypeIndex::Enum16: [[fallthrough]];
+        case TypeIndex::Decimal32: [[fallthrough]];
+        case TypeIndex::Decimal64: [[fallthrough]];
+        case TypeIndex::Decimal128: [[fallthrough]];
+        case TypeIndex::UUID:
+        {
+            col->insert(convertFieldToType(right_arg_value, *left_arg_type, right_arg_type.get()));
+            break;
+        }
+        // flatten compound values:
+        case TypeIndex::Array: [[fallthrough]];
+        case TypeIndex::Tuple: [[fallthrough]];
```

Comment:
If `left_tuple_depth == right_tuple_depth`  (case `SELECT (1, 2) IN (1, 3)`) we should interpret tuple as single value, not like array. It's also neccesary for similar case with arrays (`SELECT [1, 2] IN [1, 3]`).